### PR TITLE
Authentication User with Devise

### PR DESCRIPTION
**Pull request by IvanDanyliv**
**CHECK LIST**
- [ ] specs added
- [x] specs passed
- [ ] code coverage >90%
- [x] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [ ] "let!" is used in specs
- [x] screenshot of feature is added if task has UI

image
image

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -27,6 +27,8 @@ gem 'slim-rails', '~> 3.2.0'
 # gem 'redis', '~> 4.0'
 # Use Active Model has_secure_password
 # gem 'bcrypt', '~> 3.1.7'
+gem 'pry', '~> 0.13.1'
```
Comment (boikonataliii) [inline Gemfile:30]: move to development

File: app/views/devise/sessions/new.html.slim [hunk 2]
```diff
@@ -0,0 +1,33 @@
+head
+  title Login Page  
+body
+  = form_for(resource, as: resource_name, url: session_path(resource_name)) do |f|
+    .container
+      .d-flex.justify-content-center.h-100
+        .card
+          .card-header
+            h3 Sign In
```
Comment (mmarusyk) [inline app/views/devise/sessions/new.html.slim:9]:
```suggestion
            h3
              |Sign In
```

File: app/models/user.rb [hunk 3]
```diff
@@ -1,4 +1,9 @@
 class User < ApplicationRecord
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i.freeze
```
Comment (boikonataliii) [inline app/models/user.rb:6]: Where is this constant used?

File: app/views/devise/sessions/new.html.slim [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+              - if devise_mapping.rememberable?
+                .field
+                  = f.check_box :remember_me
+                  = f.label :remember_me
+              .actions
+                = f.submit "Log in"
+          .card-footer
+            .d-flex.justify-content-center.links
+              | Don't have an account?
+            = render "devise/shared/links"
```
Comment (mmarusyk) [inline app/views/devise/sessions/new.html.slim:33]: add new line at end of file

File: app/views/shared/_header.html.slim [hunk 5]
```diff
@@ -12,10 +12,10 @@ header.header.w-100
               span.sr-only (current)
     ul.nav.navbar-nav.navbar-right
       li
-        a.nav-link href="#" 
+        = link_to(new_user_session_path) 
           span
             button.btn.btn-transparent.auth-btn Login
       li
-        a.nav-link href="#"
+        = link_to(new_user_registration_path)
```
Comment (mmarusyk) [inline app/views/shared/_header.html.slim:19]: We can omit brackets. I think in will be better

File: app/models/user.rb (hunk 3 above)
Comment (IvanDanyliv) [reply app/models/user.rb:6]: VALID_EMAIL_REGEX це не моя таска

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -1,2 +1,9 @@
 class ApplicationController < ActionController::Base
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  protected
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:first_name, :last_name])
```
Comment (mmarusyk) [inline app/controllers/application_controller.rb:7]:
```suggestion
    devise_parameter_sanitizer.permit(:sign_up, keys:  %i[first_name last_name])
```
Use %i

File: app/models/user.rb (hunk 3 above)
Comment (boikonataliii) [reply app/models/user.rb:6]: but it looks like you added this line

File: app/models/user.rb [hunk 7]
```diff
@@ -12,22 +17,27 @@ def full_name
   private
 
   def downcase_email
```
Comment (chu4) [inline app/models/user.rb:19]: you have this in device settings
```
	  # Configure which authentication keys should be case-insensitive.
  # These keys will be downcased upon creating or modifying a user and when used
  # to authenticate or find a user. Default is :email.
  config.case_insensitive_keys = [:email]
```

File: app/views/devise/shared/_links.html.slim [hunk 8]
```diff
@@ -0,0 +1,19 @@
+- if controller_name != 'sessions'
+  = link_to "Log in", new_session_path(resource_name)
+  br
+- if devise_mapping.registerable? && controller_name != 'registrations'
+  = link_to "Sign up", new_registration_path(resource_name)
+  br
+- if devise_mapping.recoverable? && controller_name != 'passwords' && controller_name != 'registrations'
+  = link_to "Forgot your password?", new_password_path(resource_name)
+  br
+- if devise_mapping.confirmable? && controller_name != 'confirmations'
```
Comment (chu4) [inline app/views/devise/shared/_links.html.slim:10]: you dont have device settings for most of these
either add those or remove for now

File: app/models/user.rb (hunk 3 above)
Comment (chu4) [reply app/models/user.rb:6]: probably rebase troubles, remove it not used anywhere

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -1,2 +1,9 @@
 class ApplicationController < ActionController::Base
+  before_action :configure_permitted_parameters, if: :devise_controller?
+
+  protected
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: %i[first_name last_name])
+  end
```
Comment (vasya539) [inline app/controllers/application_controller.rb:8]: image
image

File: app/views/devise/sessions/new.html.slim [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
```
Comment (Ukhanskyi) [inline app/views/devise/sessions/new.html.slim:35]: rm white spaces in 35 line

**Review by mmarusyk — APPROVED** [review]

**Review by Andrii-Kozak — APPROVED** [review]

**Review by matsevich — APPROVED** [review]

**Pull request merged**